platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/uptime/UptimeTracker.java

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.uptime;
+
+import static com.swirlds.common.system.UptimeData.NO_ROUND;
+import static com.swirlds.common.units.TimeUnit.UNIT_MICROSECONDS;
+import static com.swirlds.common.units.TimeUnit.UNIT_MILLISECONDS;
+import static com.swirlds.common.units.TimeUnit.UNIT_NANOSECONDS;
+import static com.swirlds.common.units.TimeUnit.UNIT_SECONDS;
+
+import com.swirlds.common.context.PlatformContext;
+import com.swirlds.common.system.Round;
+import com.swirlds.common.system.address.Address;
+import com.swirlds.common.system.address.AddressBook;
+import com.swirlds.common.system.events.ConsensusEvent;
+import com.swirlds.common.time.Time;
+import com.swirlds.common.utility.CompareTo;
+import com.swirlds.platform.internal.ConsensusRound;
+import com.swirlds.platform.internal.EventImpl;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Monitors the uptime of nodes in the network.
+ */
+public class UptimeTracker { // TODO test
+
+    private final long selfId;
+    private final Time time;
+    private final AddressBook addressBook;
+
+    private final UptimeMetrics uptimeMetrics;
+    private final Duration degradationThreshold;
+
+    private final AtomicReference<Instant> lastEventTime = new AtomicReference<>();
+
+    /**
+     * Construct a new uptime detector.
+     *
+     * @param platformContext the platform context
+     * @param time            the time
+     * @param addressBook     the address book
+     */
+    public UptimeTracker(
+            @NonNull PlatformContext platformContext,
+            @NonNull final AddressBook addressBook,
+            final long selfId,
+            @NonNull final Time time) {
+
+        this.selfId = selfId;
+        this.time = Objects.requireNonNull(time);
+        this.addressBook = addressBook;
```

Comment:
`Objects.requireNonNull` on `addressBook` since it is not dereferenced.

Reply:
added